[docs]: add configuration reference (#397)
hyperledger-iroha/iroha-2-docs: src/api/config/sumeragi-params.md

# Sumeragi Parameters

Explain sumeragi module

## `sumeragi.trusted_peers`

::: warning Discussion

Hey, is sumeragi.trusted_peers actually required to be set in the config?

> I guess that's the only way to specify the relevant topology and let the nodes know what addresses to communicate with as well as what public keys to use to verify respective node signatures

Reviewer: No guessing, either yes or no please.

Author: This text is not for the users... it is for us to discuss.

# Sumeragi Parameters

TODO Explain sumeragi module. Clarify that it consists of transactions and blocks gossiping, which are two separate
processes. Explain how periods & gossip sizes might affect network load & speed of sync. Explain the pipeline of
transactions and blocks, refer to consensus section and maybe some others.

## `sumeragi.block_time`

- **Type:** String or Number, [Duration](glossary#type-duration)
- **Default:** 2 seconds

Time since the round start[^1] by which a new block should be created regardless if there were enough transactions or
not. Used to force block commits when there is a small influx of new transactions.

A block might be created earlier if there is enough transactions in the [Queue](queue-params). The limit of transactions
is configured by [`sumeragi.transactions_in_block`](#sumeragi-transactions-in-block).

**Example:**

```toml
[sumeragi]
block_time = "2s"
```

## `sumeragi.block_gossip_period`

- **Type:** String or Number, [Duration](glossary#type-duration)
- **Default:** 10 seconds

> The time interval between requests to peers for the most recent block.

Reviewer 2: `requests to peers for the most recent block` does this mean requests *about* the most recent block?

Author: Yes. Request the most recent block from peers.

Reviewer 2: then maybe

```suggestion
The time interval between sending requests to peers about the most recent block.
```

**Example:**

```toml
[sumeragi]
block_gossip_period = "5s"
```

::: warning

More frequent gossiping shortens the time to sync, but can overload the network.

:::

## `sumeragi.commit_time`

- **Type:** String or Number, [Duration](glossary#type-duration)
- **Default:** 4 seconds

Time by which a newly created block should be committed. Prevents malicious nodes from stalling the network by not
participating in consensus.

The block creation is configured with [`sumeragi.block_time`](#sumeragi-block-time) and
[`sumeragi.transactions_in_block`](#sumeragi-transactions-in-block).

**Example:**

```toml
[sumeragi]
commit_time = "4s"
```

## `sumeragi.max_blocks_per_gossip`

- **Type:** Number
- **Default:** $4$

> The amount of blocks that can be sent in a single synchronization message.

Reviewer 2: ```suggestion
The number of blocks that can be sent in a single synchronization message.
```

**Example:**

```toml
[sumeragi]
max_blocks_per_gossip = 4
```

## `sumeragi.max_transactions_per_gossip`

- **Type:** Number
- **Default:** $500$

> Max number of transactions in a gossip batch message. Smaller size leads to longer time to synchronise, but useful if

Reviewer 2: ```suggestion
Max number of transactions in a gossip batch message. Smaller size leads to longer time to synchronise, but can be useful if
```

you have high packet loss.

**Example:**

```toml
[sumeragi]
max_transactions_per_gossip = 500
```

## `sumeragi.transaction_gossip_period`

- **Type:** String or Number, [Duration](glossary#type-duration)
- **Default:** 1 second

Period for pending transaction gossiping between peers.

**Example:**

```toml
[sumeragi]
transaction_gossip_period = "1s"
```

::: warning

> More frequent gossiping shortens the time to sync, but can overload the network.

Reviewer 2: ```suggestion
More frequent gossiping shortens the time to sync but can overload the network.
```

:::

## `sumeragi.transactions_in_block`

- **Type:** u32
- **Default:** $2^9 = 512$

The upper limit of the number of transactions per block. If there is enough transactions in the [Queue](queue-params),
the block is created immediately. Otherwise, the block is created when [`sumeragi.block_time`](#sumeragi-block-time) is

> elapsed since the round start[^1].

Reviewer 2: ```suggestion
elapsed since the start of the round[^1].
```

  - **Required**
- **`public_key`:** Public key of the peer.
  - **Type:** String, [Multi-hash](glossary#type-multi-hash)
  - **Required**

```toml
peer_id = { address = "localhost:1338", public_key = "ed0120FAFCB2B27444221717F6FCBF900D5BE95273B1B0904B08C736B32A19F16AC1F9" }
```

An array can be defined in two ways. Like this:

```toml
[[sumeragi.trusted_peers]]
address = "localhost:1338"
public_key = "ed012067C02E340AADD553BCF7DB28DD1F3BE8BE3D7581A2BAD81580AEE5CC75FEBD45"

[[sumeragi.trusted_peers]]
address = "localhost:1339"
public_key = "ed0120236808A6D4C12C91CA19E54686C2B8F5F3A786278E3824B4571EF234DEC8683B"

[[sumeragi.trusted_peers]]
address = "localhost:1340"
public_key = "ed0120FAFCB2B27444221717F6FCBF900D5BE95273B1B0904B08C736B32A19F16AC1F9"
```

Or like this:

```toml
[sumeragi]
trusted_peers = [
  { address = "localhost:1338", public_key = "ed012067C02E340AADD553BCF7DB28DD1F3BE8BE3D7581A2BAD81580AEE5CC75FEBD45" },
  { address = "localhost:1339", public_key = "ed0120236808A6D4C12C91CA19E54686C2B8F5F3A786278E3824B4571EF234DEC8683B" },
  { address = "localhost:1340", public_key = "ed0120FAFCB2B27444221717F6FCBF900D5BE95273B1B0904B08C736B32A19F16AC1F9" },
]
```

[^1]:

>     The round start happens on peers rotation, when the leader is elected. Generally it happens after the previous block
>     is committed. See [Consensus](/guide/blockchain/consensus) (todo: that page doesn't mention "round start" term ).

Reviewer 2: ```suggestion
    The start of the round happens when the leader is elected during peer rotation. Generally it happens after the previous block
    is committed. See [Consensus](/guide/blockchain/consensus) (todo: that page doesn't mention "round start" term ).
```
